assignment3/tictactoe.py

```diff
@@ -0,0 +1,166 @@
+        Test your program by playing a few games.
+
+    On assembling this program, the assignment author found that it was too time consuming to write some of the methods.  So, here are some pieces to reuse.  Please make sure you understand them.
+    """
+
+    valid_moves = [
+        "upper left",
+        "upper center",
+        "upper right",
+        "middle left",
+        "center",
+        "middle right",
+        "lower left",
+        "lower center",
+        "lower right",
+    ]
+
+    def __init__(self):
+        self.board_array = [
+            [" ", " ", " "],
+            [" ", " ", " "],
+            [" ", " ", " "],
+        ]
+        self.last_move = " "  # used to access the last player moving
+        self.total_turns = 1  # each player's turn increments this number
+
+    def move(self, move_string):
+        if move_string not in self.valid_moves:
+            raise TictactoeException("That's not a valid move.")
+
+        column = self.valid_moves.index(move_string) % 3
+        row = self.valid_moves.index(move_string) // 3
+
+        new_row = self.board_array[row]
+
+        if new_row[column] != " ":
+            raise TictactoeException("That spot is taken.")
+
+        new_row[column] = self.turn
+        self.board_array[row] = new_row
+        self.next_turn()
+
+    @property
+    def turn(self):
+        return ["o", "x"][self.total_turns % 2]
+
+    def next_turn(self):
+        self.last_move = ["o", "x"][self.total_turns % 2]
+        self.total_turns += 1
+        return ["o", "x"][self.total_turns % 2]
+
+    def whats_next(self):
+        for row in self.board_array:
+            if "".join(row) in ["xxx", "ooo"]:
+                return (True, row[0])
+
+        for column in range(3):
+            if (
+                self.board_array[0][column]
+                == self.board_array[1][column]
+                == self.board_array[2][column]
+                != " "
+            ):
+                return (True, self.board_array[0][column])
+
+        if (
+            self.board_array[0][0]
+            == self.board_array[1][1]
+            == self.board_array[2][2]
+            != " "
+        ):
+            return (True, self.board_array[0][0])
+        if (
+            self.board_array[0][2]
+            == self.board_array[1][1]
+            == self.board_array[2][0]
+            != " "
+        ):
+            return (True, self.board_array[0][2])
+
+        board_str = "".join(move for row in self.board_array for move in row)
+        if not board_str.count(" "):
+            # The game has ended. Its a draw.
+            return (True, "Cat's Game")
+        return (False, self.turn)  # Game hasn't ended.
```

Comment:
Again, I suggest you don't have constructs with multiple return values. I instead recommend something like message="X has won!" or whatever, and game_over=True (or false), and then at the end:
return (game_over, message)

It is hard to understand the current construct with return (Bool, self.board_array[x][y]). What is the intent? What is being returned and why? When you have explicitly defined variables (game_over, message (or whatever), and then return them at the end, it makes the code self-commenting and much easier to follow.